tree/30_generic_methods/user_group.cf

```diff
@@ -0,0 +1,134 @@
+#####################################################################################
+# Copyright 2020 Normation SAS
+#####################################################################################
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, Version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name User group
+# @description Define secondary group for a user
+# @documentation Ensure that a user is within a group
+#
+# #### Behavour
+#
+# Ensure that the user belongs in the given secondary group (non-exclusive)
+#
+# ##### Parameters
+#
+# `user`      : the user login
+# `group_name`: secondary group name the user should belong to (non-exclusive)
+#
+# #### Examples
+#
+# To ensure that user `test` belongs in group `dev`
+#
+# ```
+#  user_group("test", "99_rudder", "0", "")
+# ```
+# Note that it will make sure that user test is in group dev, but won't remove it
+# from other groups it may belong to
+#
+# @parameter user           User login
+# @parameter group_name     Secondary group name for the user
+# @class_prefix user_group
+# @class_parameter user
+
+
+
+bundle agent user_group(user, group_name) {
+  vars:
+      "old_class_prefix"      string => canonify("user_group_${user}");
+
+      "args"                   slist => { "${user}", "${group_name}" };
+      "report_param"          string => join("_", args);
+      "full_class_prefix"     string => canonify("user_group_${report_param}");
+      "class_prefix"          string => string_head("${full_class_prefix}", "1000");
+
+      "add_user_to_group_cmd"   string => "${paths.usermod} -a -G ${group_name} ${user}";
+      "full_inner_class_prefix" string => canonify("command_execution_${add_user_to_group_cmd}");
+      "inner_class_prefix"      string => string_head("${full_inner_class_prefix}", "1000");
+
+
+    user_exists.group_exists::
+      "groups_test"           string => execresult("${ncf_paths.path[id]} -Gn ${user}", "noshell");
+      "current_groups"         slist => string_split("${groups_test}", " ", "999999");
+
+    # if we needed to change the group, check groups afterwards
+    pass2.!within_group::
+      "groups_test_post_change"   string => execresult("${ncf_paths.path[id]} -Gn ${user} #post change", "useshell"),
```

Comment:
Is this really re-evaluated each time we call the bundle? I think that if you nedd to call it twice in a run, it will note execute the second one. Perhaps i am confusing with the returnszero function.

Reply:
no, it's not reevaluted, but I need to run it only only if it's not within_group. As within_group is defined in pass1 of class, it needs to be run in pass2 of variable